### PR TITLE
Add http proxy support for exec/port-forward

### PR DESCRIPTION
**Pull request by ncdc**
Add http proxy support for exec/port-forward in SpdyRoundTripper

### CODE REVIEW

**Comment by ncdc** [comment]
@smarterclayton

File: pkg/util/httpstream/spdy/roundtripper.go [hunk 1]
```diff
@@ -50,21 +52,82 @@ type SpdyRoundTripper struct {
...
+		return nil, err
+	}
+
+	if proxyURL == nil {
+		return s.dialWithoutProxy(req.URL)
+	}
+
+	// proxying logic adapted from http://blog.h6t.eu/post/74098062923/golang-websocket-with-http-proxy-support
+	proxyReq := http.Request{
+		Method: "CONNECT",
```
Comment (ncdc) [inline pkg/util/httpstream/spdy/roundtripper.go]: I'm not sure if it's appropriate to do a `CONNECT` tunnel for regular http traffic. If not, what's the right way to do it?

Comment (smarterclayton) [inline pkg/util/httpstream/spdy/roundtripper.go]: Doing a quick read I think so

**Comment by smarterclayton** [comment]
Needs a unit test somehow... no idea how to do that easily though.

**Comment by smarterclayton** [comment]
http://www.infoq.com/articles/Web-Sockets-Proxy-Servers seems to cover this for WS

**Comment by ncdc** [comment]
@smarterclayton I'm not sure how you'd unit test it, since everything would be running on the same host. But I agree some sort of way to test is highly desirable.

**Comment by smarterclayton** [comment]
You'd need a connect level proxy - which you could run in a container
during e2e

On Fri, Jul 24, 2015 at 1:18 PM, Andy Goldstein <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> I'm not sure how
> you'd unit test it, since everything would be running on the same host. But
> I agree some sort of way to test is highly desirable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11694#issuecomment-124584396>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by ncdc** [comment]
I'm thinking that the way to e2e test this is to do the following:

1. Create a pod with a proxy server such as squid in it
1. Create a pod with kubectl in it
1. Configure the kubectl container such that it's only allowed to talk to the squid pod (or a service fronting it)
1. Have the kubectl container run `http_proxy=$squid kubectl -s $apiServer exec $somePod $someCommand`

Questions:

1. Do you think this will work? Is there an easier way?
1. Can we use iptables to restrict the kubectl container's outgoing packets?
  1. Is it safe to assume iptables will work inside a container wherever e2e is running?
  1. This means the kubectl container will need to be privileged - is that ok?
  1. Is there some other way to do this?
1. Ideally I'd like to test `http_proxy` and `https_proxy` - can we test both http and https with e2e?

@smarterclayton @thockin @bgrant0607 @brendandburns

**Comment by ncdc** [comment]
@kubernetes/rh-cluster-infra @kubernetes/rh-networking

**Comment by bgrant0607** [comment]
cc @lavalamp

**Comment by bgrant0607** [comment]
Why do you want to restrict packets?

**Comment by ncdc** [comment]
I'd like to prove without a doubt that proxying is working. If we can't restrict packets, I could write a test that runs `http_proxy=... kubectl exec ...` and succeeds with the current code, which doesn't support proxying.

**Comment by thockin** [comment]
regarding iptables - it should work inside containers, but it does require
privileges.

**Comment by ncdc** [comment]
> regarding iptables - it should work inside containers, but it does require
privileges.

Yes, it seems somewhat backwards to require a privileged container to be able to test that proxying is working, but that's all I've been able to come up with so far...

**Comment by ncdc** [comment]
@mvdan my fear is this will never be called https://github.com/kubernetes/kubernetes/pull/13058/files#diff-3c23f834eaf7e756813f19c5a1116c46R295

**Comment by mvdan** [comment]
@ncdc could you elaborate? The intended behaviour of the code in portforward is the same, with the only difference that the old one was broken.

**Comment by ncdc** [comment]
@mvdan sorry that comment was meant for the go1.5 race pr

**Comment by lavalamp** [comment]
Why not just examine the kubectl log to verify that all traffic went through its proxy?

**Comment by ncdc** [comment]
@lavalamp good idea

**Comment by ncdc** [comment]
Do you all have a preference for what image to use for the proxy? Something with squid or polipo in it?

**Comment by ncdc** [comment]
Actually I'll try nginx since I know e2e already uses that

**Comment by ncdc** [comment]
Looks like nginx doesn't support CONNECT when acting as a forward proxy. That leaves me with squid, polipo, ...

**Comment by ncdc** [comment]
If we run the forward proxy as a pod, and we run `kubectl exec` from outside the cluster, there's no easy way for kubectl to use the proxy pod as its http_proxy, since the pod's IP isn't routable from outside the cluster.

We could try to find a way to run kubectl inside the cluster, but that requires that kubectl be built locally and then somehow run in the cluster, perhaps in a custom image that's built on the fly, pushed to a publicly accessible registry, and then run as a pod. I don't like that idea because of the need to build and push a kubectl image somewhere.

Another option would be to run the proxy pod with a hostPort, perhaps, and have kubectl connect to that.

Does anyone have any thoughts about how to put together an e2e test where kubectl uses an http proxy to talk to the apiserver?

**Comment by smarterclayton** [comment]
Run kubectl in a pod

On Aug 27, 2015, at 5:26 PM, Andy Goldstein <notifications@github.com>
wrote:

If we run the forward proxy as a pod, and we run kubectl exec from outside
the cluster, there's no easy way for kubectl to use the proxy pod as its
http_proxy, since the pod's IP isn't routable from outside the cluster.

We could try to find a way to run kubectl inside the cluster, but that
requires that kubectl be built locally and then somehow run in the cluster,
perhaps in a custom image that's built on the fly, pushed to a publicly
accessible registry, and then run as a pod. I don't like that idea because
of the need to build and push a kubectl image somewhere.

Another option would be to run the proxy pod with a hostPort, perhaps, and
have kubectl connect to that.

Does anyone have any thoughts about how to put together an e2e test where
kubectl uses an http proxy to talk to the apiserver?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/11694#issuecomment-135544986>
.

**Comment by lavalamp** [comment]
Yeah, I agree with Clayton. We need to get the workflow of "building binary
-> building container -> build pod definition -> run in test cluster"
working.

**Comment by ncdc** [comment]
Do you have any thoughts on how to get a locally built kubectl running in a
pod? I had an idea of uploading the binary somewhere and then running a pod
whose container would download the file and then run it. That seems better
than always building and pushing an image somewhere.

**Comment by smarterclayton** [comment]
From the e2e test you could start a container and exec in the binary. Other
ways might be a hostPath mount. You *might* be able to abuse a container
that expects a stdin stream on input it writes to disk and then executes.

On Aug 27, 2015, at 9:57 PM, Andy Goldstein <notifications@github.com>
wrote:

Do you have any thoughts on how to get a locally built kubectl running in a
pod? I had an idea of uploading the binary somewhere and then running a pod
whose container would download the file and then run it. That seems better
than always building and pushing an image somewhere.
On Thu, Aug 27, 2015 at 8:53 PM Daniel Smith <notifications@github.com>
wrote:

> Yeah, I agree with Clayton. We need to get the workflow of "building
binary
> -> building container -> build pod definition -> run in test cluster"
> working.
>
> On Thu, Aug 27, 2015 at 2:49 PM, Clayton Coleman <notifications@github.com
> >
> wrote:
>
> > Run kubectl in a pod
> >
> > On Aug 27, 2015, at 5:26 PM, Andy Goldstein <notifications@github.com>
> > wrote:
> >
> > If we run the forward proxy as a pod, and we run kubectl exec from
> outside
> > the cluster, there's no easy way for kubectl to use the proxy pod as its
> > http_proxy, since the pod's IP isn't routable from outside the cluster.
> >
> > We could try to find a way to run kubectl inside the cluster, but that
> > requires that kubectl be built locally and then somehow run in the
> cluster,
> > perhaps in a custom image that's built on the fly, pushed to a publicly
> > accessible registry, and then run as a pod. I don't like that idea
> because
> > of the need to build and push a kubectl image somewhere.
> >
> > Another option would be to run the proxy pod with a hostPort, perhaps,
> and
> > have kubectl connect to that.
> >
> > Does anyone have any thoughts about how to put together an e2e test
where
> > kubectl uses an http proxy to talk to the apiserver?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> >
> https://github.com/kubernetes/kubernetes/pull/11694#issuecomment-135544986
> > >
> > .
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/kubernetes/kubernetes/pull/11694#issuecomment-135565422
> >
> > .
> >
>
> —
> Reply to this email directly or view it on GitHub
> <
https://github.com/kubernetes/kubernetes/pull/11694#issuecomment-135593944>
> .
>

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/11694#issuecomment-135594413>
.

**Comment by lavalamp** [comment]
Yeah-- you can build kubectl into a container right now (there's a Dockerfile in the repo to do it) but the issue is that there's no general way to transfer the built container into the cluster. We need a private-repo service; then it'd be trivial.

Until someone writes that component, you could make a general "exec this blurb" pod and transfer in the freshly built kubectl, like Clayton suggests.

**Comment by ncdc** [comment]
I'm working on the exec flow, but I have to fix some exec bugs first :-(

**Comment by ncdc** [comment]
... still working on e2e for this. #13322 is a prereq to be able to run `rsync -e 'kubectl exec ...' ...` assuming that's the way I end up going to inject the binary into the target test container. Another option would be to create an image that runs a simple http server that handles a file upload and writes it to the container's file system, after which we could run the binary.

**Comment by ncdc** [comment]
@ashcrow along with rebasing, you may need to look into this failure on GCE:

```
• Failure [216.311 seconds]
Kubectl client
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:764
  Simple pod
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:335
    should support exec through an HTTP proxy [It]
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:288

    Sep 25 09:45:27.222: No token found for e2e user. Output: 

    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:209
```

**Comment by ncdc** [comment]
Shippable error:

```
cp: cannot stat `/root/workspace/src/github.com/kubernetes/kubernetes/shippable': No such file or directory
test/images/goproxy/goproxy.go:23:2: cannot find package "github.com/elazarl/goproxy" in any of:
	/root/.gvm/gos/go1.4/src/github.com/elazarl/goproxy (from $GOROOT)
	/root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/elazarl/goproxy (from $GOPATH)
	/root/workspace/src/github.com/elazarl/goproxy
godep: go exit status 1
```

**Comment by ncdc** [comment]
e2e flake

```
ERROR: (gcloud.compute.addresses.create) Some requests did not succeed:
 - Quota 'STATIC_ADDRESSES' exceeded.  Limit: 25.0
```

@k8s-bot test this please

**Comment by ashcrow** [comment]
Same flake

```
ERROR: (gcloud.compute.addresses.create) Some requests did not succeed:
 - Quota 'STATIC_ADDRESSES' exceeded.  Limit: 25.0
```

**Comment by ashcrow** [comment]
Flake changed :smile:

```
There are no objects in this folder.
```

**Comment by ixdy** [comment]
ugh, having issues with our infrastructure. let's try again.
@k8s-bot this this

**Comment by ashcrow** [comment]
@k8s-bot test this please

**Comment by ncdc** [comment]
@k8s-bot test this please

**Comment by ashcrow** [comment]
Latest failure is:

```
Sep 28 13:30:13.784: <snip> the path "/jenkins-master-data/jobs/kubernetes-pull-build-test-e2e-gce/workspace@5/kubernetes/contrib/for-tests/netexec/pod.yaml" does not exist
```

That's progress. However, the file [should exist](https://github.com/ncdc/kubernetes/blob/6b753e305d095c1f9f602775f56398948240eb48/contrib/for-tests/netexec/pod.yaml) in the checkout. 

Is ```/contrib``` stripped out of the checkouts by chance?

**Comment by ixdy** [comment]
I think the issue is that nothing under `contrib/for-tests` is included in the kubernetes-test tarball, so those dependencies aren't available to the e2e tests.

**Comment by ncdc** [comment]
It looks like the other images moved to test/images - perhaps we could move netexec there?

**Comment by ashcrow** [comment]
I'll move netexec over to ```test/images``` in the netexec commit and give it a go.

**Comment by ashcrow** [comment]
Hrm, is ```test/images``` also not included?
```
    the path "/jenkins-master-data/jobs/kubernetes-pull-build-test-e2e-gce/workspace@3/kubernetes/test/images/netexec/pod.yaml" does not exist
```

**Comment by ncdc** [comment]
It passed!

File: hack/build-cross.sh [hunk 2]
```diff
@@ -28,6 +28,7 @@ KUBE_BUILD_PLATFORMS=("${KUBE_SERVER_PLATFORMS[@]}")
 kube::golang::build_binaries "${KUBE_SERVER_TARGETS[@]}"
 
 KUBE_BUILD_PLATFORMS=("${KUBE_CLIENT_PLATFORMS[@]}")
+KUBE_STATIC_OVERRIDES=("kubectl")
```
Comment (ncdc) [inline hack/build-cross.sh]: @smarterclayton @mikedanese is this ok? We need a static kubectl so we can run it in a container as part of the e2e tests for exec + http proxy.

Comment (smarterclayton) [inline hack/build-cross.sh]: What's the static kubectl downside?  Can't get the current user and can't
use libc DNS?

**Comment by ashcrow** [comment]
The travis and shippable error may be related to https://github.com/kubernetes/kubernetes/issues/14677

```
package github.com/google/google-api-go-client/cloudmonitoring/v2beta2
	imports github.com/google/google-api-go-client/cloudmonitoring/v2beta2
	imports github.com/google/google-api-go-client/cloudmonitoring/v2beta2: code in directory /tmp/gopath.JGnmyy/src/github.com/google/google-api-go-client/cloudmonitoring/v2beta2 expects import "google.golang.org/api/cloudmonitoring/v2beta2
```

**Comment by ncdc** [comment]
They are. We probably need a separate PR to fix that. It's been in since 9/17, though...

**Comment by ncdc** [comment]
Or rather, they're verify-godeps.sh related, but a different root cause from 14677.

**Comment by ncdc** [comment]
https://github.com/kubernetes/kubernetes/pull/14770

File: hack/build-cross.sh (hunk 2 above)
Comment (ncdc) [inline hack/build-cross.sh]: Probably. I'm hesitant to force a static build of kubectl just to satisfy a new e2e test for exec+proxy. Maybe we could find a way to produce both static and dynamic builds (e.g. kubectl remains dynamic, add kubectl-static)?

cc @brendandburns @bgrant0607 @lavalamp

**Comment by ncdc** [comment]
@smarterclayton we finally got all tests to pass everywhere

**Comment by ncdc** [comment]
I think we still need to resolve the question about changing kubectl to be built static 100% of the time.

File: hack/build-cross.sh (hunk 2 above)
Comment (ashcrow) [inline hack/build-cross.sh]: It's possible. However, I'd like to find out if there is a requirement for dynamic first. If either is OK I'd say running with static would make sense. If there is a requirement for a dynamic ```kubectl``` along with our new need for a static one then we probably need to update build scripts in such a way that they produce both.

Comment (ncdc) [inline hack/build-cross.sh]: Agreed. Let's find out if dynamic is required or not.

Comment (pmorie) [inline hack/build-cross.sh]: I have the same question as clayton re: downside.  Static would be good unless we want to rethink how the images of this test are going to be structured.

**Comment by jlowdermilk** [comment]
Re static kubectl: inability to get current user shouldn't be a problem. The loader uses env `$HOME` to determine default config dir location. See [clientcmd/loader.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/clientcmd/loader.go#L42), #14087. Not sure about the libc dns question.

**Comment by ashcrow** [comment]
@jlowdermilk can you think of any reason we'd need to specify the use of libc dns or not?

**Comment by jlowdermilk** [comment]
@ashcrow, don't know. I don't think we use it in pkg/unversioned/client or its dependencies, but @brendandburns or @lavalamp may know otherwise.

**Comment by ncdc** [comment]
@smarterclayton are you able to review this, or should we reassign?

**Comment by smarterclayton** [comment]
Reassign, traveling

On Oct 1, 2015, at 9:03 PM, Andy Goldstein <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> are you able to review
this, or should we reassign?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/11694#issuecomment-144817823>
.

**Comment by ncdc** [comment]
@krousey do you have time to review this?

**Comment by ncdc** [comment]
Note, this PR adds 1 new image and updates an existing image, so those will need to be pushed and the PR updated accordingly before it is merged.

**Comment by krousey** [comment]
@ncdc probably not today. I could look at it tomorrow.

**Comment by ncdc** [comment]
@kubernetes/rh-cluster-infra PTAL if you have time to review

**Comment by krousey** [comment]
I looked through it with no major comments. So I assume you want me to build and deploy netexec and goproxy?

**Comment by ncdc** [comment]
@krousey yes, please!

**Comment by krousey** [comment]
@ncdc sorry that took so long. Wanted to get my work in order before checking out your branch. Both images have been uploaded.

**Comment by ashcrow** [comment]
@krousey thanks for uploading them! Updated files to point to gcr.io.

File: test/images/goproxy/Makefile [hunk 3]
```diff
@@ -0,0 +1,15 @@
+all: push
+
+TAG = 0.1
+
+goproxy: goproxy.go
+	CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -ldflags '-w' ./goproxy.go
+
+image: goproxy
+	sudo docker build -t gcr.io/google_containers/goproxy:$(TAG) .
```
Comment (krousey) [inline test/images/goproxy/Makefile]: Please remove this sudo

File: hack/build-cross.sh (hunk 2 above)
Comment (krousey) [inline hack/build-cross.sh]: Has this issue been resolved? Or should this PR wait on this?

File: test/images/goproxy/Makefile (hunk 3 above)
Comment (ashcrow) [inline test/images/goproxy/Makefile]: Done.

File: hack/build-cross.sh (hunk 2 above)
Comment (ashcrow) [inline hack/build-cross.sh]: @krousey no one has chimed in why a downside but I wouldn't consider it resolved. Though I'm not sure who to tag who could answer the question.

Comment (lavalamp) [inline hack/build-cross.sh]: There's a makefile around somewhere that does this specifically for the container build (and also builds the container). I don't think we need to make the default build static, do we?

It's probably in contrib?

Comment (ashcrow) [inline hack/build-cross.sh]: Looks like ```/examples/kubectl-container/Makefile``` sets ```KUBE_STATIC_OVERRIDES``` to ```"kubectl"```.

@lavalamp how would you recommend executing and referencing it in jenkins e2e? As part of the e2e test itself before uploading into ```netexec```?

**Comment by ashcrow** [comment]
Flake. Rebuilding.

File: hack/build-cross.sh (hunk 2 above)
Comment (lavalamp) [inline hack/build-cross.sh]: That part is not solved yet, which is extremely unfortunate. "rebuild &
push periodically" is the best answer I have right now. We need to run a
private registry in e2e clusters so there's a place to push & pull these
random containers to/from.

Comment (ashcrow) [inline hack/build-cross.sh]: OK. Soon I'll look at simply running the build and publishing the static result up to the ```netexec``` container as part of the the test run. I'll then remove the forced static ```kubectl``` from the main build.

**Comment by krousey** [comment]
@ashcrow @ncdc Coming back to this after a brief vacation, is the static kubectl issue fixed?

**Comment by krousey** [comment]
@k8s-bot test this

**Comment by ncdc** [comment]
@krousey not fixed yet, @ashcrow is working on something else at the moment, but we'll come back to it soon.

**Comment by krousey** [comment]
@ncdc Thanks. I mainly wanted to get that status in the main comment stream so that I don't get confused again and again :smile:

**Comment by ashcrow** [comment]
@krousey jumping back over now :smile:

**Comment by ashcrow** [comment]
@k8s-teamcity-mesosphere failure *seems* like a flake

```
[13:54:14][Step 4/4] Error response from daemon: Cannot start container 49b2e6b7cb245420fc0b4b270842a38f781329fabbdb123d1afcac79b12de43c: Cannot link to a non running container: /docker_apiserver_1 AS /prickly_bose/apiserver
[13:54:14][Step 4/4] Dumping logs to '/tmp/kubernetes/log'
[13:54:15][Step 4/4] + timeout 5m ./cluster/kube-down.sh
[13:54:15][Step 4/4]  [1BBringing down cluster using provider: mesos/docker
[13:54:15][Step 4/4] Verifying required commands
[13:54:15][Step 4/4] Stopping mesos/docker cluster
```

**Comment by ashcrow** [comment]
Shippable also looks like a flake. One of them passed in 40 minutes and the other failed in 2 hours.

**Comment by ashcrow** [comment]
@krousey mind taking a look now that this no longer forces a static kubectl for everything?

**Comment by ashcrow** [comment]
Unrelated Jenkins e2e error:

```
• Failure [94.077 seconds]
EmptyDir volumes
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/empty_dir.go:95
  should support (root,0644,tmpfs) [Conformance] [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/empty_dir.go:46

  Oct 14 21:49:39.061: Failed to create pod: Pod "pod-243d789a-72f8-11e5-bdf9-42010af0226b" is forbidden: no API token found for service account e2e-tests-emptydir-vjc6c/default, retry after the token is automatically created and added to the service account

  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/util.go:1112
```

**Comment by wojtek-t** [comment]
@k8s-bot test this please

**Pull request merged**